### PR TITLE
benchmarks/coremark: update SEED_METHOD/MEM_METHOD and debug

### PR DESCRIPTION
**Pull request by pengxiang-hao**
## Summary

This patch update coremark with  SEED_METHOD/MEM_METHOD/RUN_METHOD and debug which provided by @anchao 

## Impact

new SEED_METHOD&MEM_METHOD add in coremark

## Testing

Pass Vela CI

### CODE REVIEW

File: benchmarks/coremark/Makefile [hunk 1]
```diff
@@ -57,6 +57,15 @@ else
 COREMARK_PRINT_ARGS = 0
 endif
 
+CFLAGS += ${DEFINE_PREFIX}CALLGRIND_RUN=0
+CFLAGS += ${DEFINE_PREFIX}COMPILER_REQUIRES_SORT_RETURN=0
+CFLAGS += ${DEFINE_PREFIX}CORE_DEBUG=$(CONFIG_COREMARK_DEBUG)
+CFLAGS += ${DEFINE_PREFIX}MEM_METHOD=$(CONFIG_COREMARK_MEM_METHOD)
+CFLAGS += ${DEFINE_PREFIX}MICA=0
+CFLAGS += ${DEFINE_PREFIX}PRINT_ARGS=0
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Makefile:65]: remove, define at line 72

File: benchmarks/coremark/Makefile [hunk 2]
```diff
@@ -57,6 +57,15 @@ else
 COREMARK_PRINT_ARGS = 0
 endif
 
+CFLAGS += ${DEFINE_PREFIX}CALLGRIND_RUN=0
+CFLAGS += ${DEFINE_PREFIX}COMPILER_REQUIRES_SORT_RETURN=0
+CFLAGS += ${DEFINE_PREFIX}CORE_DEBUG=$(CONFIG_COREMARK_DEBUG)
+CFLAGS += ${DEFINE_PREFIX}MEM_METHOD=$(CONFIG_COREMARK_MEM_METHOD)
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Makefile:63]: MEM_LOCATION need align with MEM_METHOD

File: benchmarks/coremark/Makefile [hunk 3]
```diff
@@ -57,6 +57,15 @@ else
...
+CFLAGS += ${DEFINE_PREFIX}MICA=0
+CFLAGS += ${DEFINE_PREFIX}PRINT_ARGS=0
+CFLAGS += ${DEFINE_PREFIX}SEED_METHOD=$(CONFIG_COREMARK_SEED_METHOD)
+CFLAGS += ${DEFINE_PREFIX}USE_CLOCK=1
+CFLAGS += ${DEFINE_PREFIX}COMPILER_FLAGS="\"$(CFLAGS)\""
 CFLAGS += ${DEFINE_PREFIX}USE_PTHREAD ${DEFINE_PREFIX}PERFORMANCE_RUN=1
 CFLAGS += ${DEFINE_PREFIX}MULTITHREAD=$(COREMARK_NTHREADS)
 CFLAGS += ${DEFINE_PREFIX}FLAGS_STR=\""$(ARCHOPTIMIZATION)\""
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Makefile:71]: remove FLAGS_STR, due with COMPILER_FLAGS

File: benchmarks/coremark/Makefile [hunk 4]
```diff
@@ -57,6 +57,15 @@ else
...
+CFLAGS += ${DEFINE_PREFIX}CALLGRIND_RUN=0
+CFLAGS += ${DEFINE_PREFIX}COMPILER_REQUIRES_SORT_RETURN=0
+CFLAGS += ${DEFINE_PREFIX}CORE_DEBUG=$(CONFIG_COREMARK_DEBUG)
+CFLAGS += ${DEFINE_PREFIX}MEM_METHOD=$(CONFIG_COREMARK_MEM_METHOD)
+CFLAGS += ${DEFINE_PREFIX}MICA=0
+CFLAGS += ${DEFINE_PREFIX}PRINT_ARGS=0
+CFLAGS += ${DEFINE_PREFIX}SEED_METHOD=$(CONFIG_COREMARK_SEED_METHOD)
+CFLAGS += ${DEFINE_PREFIX}USE_CLOCK=1
+CFLAGS += ${DEFINE_PREFIX}COMPILER_FLAGS="\"$(CFLAGS)\""
 CFLAGS += ${DEFINE_PREFIX}USE_PTHREAD ${DEFINE_PREFIX}PERFORMANCE_RUN=1
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Makefile:69]: COREMARK_RUN_METHOD need change with COREMARK_RUN_METHOD

File: benchmarks/coremark/Makefile [hunk 5]
```diff
@@ -86,6 +95,7 @@ $(COREMARK_UNPACKNAME): $(COREMARK_ZIP)
 	$(Q) $(UNPACK) $(COREMARK_ZIP)
 	$(Q) mv	coremark-$(COREMARK_VERSION) $(COREMARK_UNPACKNAME)
 	$(Q) patch -p0 < add_init_message.patch # Update init message
+	$(Q) patch -p0 < fix_build_in_run.patch # Fix build break in run
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Makefile:98]: don't need xxxx_RUN could be defined in Makefile

File: benchmarks/coremark/Makefile [hunk 6]
```diff
@@ -57,14 +57,24 @@ else
...
-CFLAGS += ${DEFINE_PREFIX}PRINT_ARGS=$(COREMARK_PRINT_ARGS)
 CFLAGS += ${DEFINE_PREFIX}MEM_LOCATION=\"Stack\"
+CFLAGS += ${DEFINE_PREFIX}MEM_METHOD=$(CONFIG_COREMARK_MEM_METHOD)
+CFLAGS += ${DEFINE_PREFIX}MICA=0
+CFLAGS += ${DEFINE_PREFIX}PRINT_ARGS=0
+CFLAGS += ${DEFINE_PREFIX}MULTITHREAD=$(COREMARK_NTHREADS)
+CFLAGS += ${DEFINE_PREFIX}PRINT_ARGS=$(COREMARK_PRINT_ARGS)
+CFLAGS += ${DEFINE_PREFIX}SEED_METHOD=$(CONFIG_COREMARK_SEED_METHOD)
+CFLAGS += ${DEFINE_PREFIX}USE_CLOCK=1
+CFLAGS += ${DEFINE_PREFIX}USE_PTHREAD ${DEFINE_PREFIX}PERFORMANCE_RUN=1
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Makefile:72]:
```
CFLAGS += ${DEFINE_PREFIX}MULTITHREAD=$(COREMARK_NTHREADS)
CFLAGS += ${DEFINE_PREFIX}PERFORMANCE_RUN=1
CFLAGS += ${DEFINE_PREFIX}PRINT_ARGS=$(COREMARK_PRINT_ARGS)
CFLAGS += ${DEFINE_PREFIX}SEED_METHOD=$(CONFIG_COREMARK_SEED_METHOD)
CFLAGS += ${DEFINE_PREFIX}USE_CLOCK=1
CFLAGS += ${DEFINE_PREFIX}USE_PTHREAD 
```

File: benchmarks/coremark/Makefile [hunk 7]
```diff
@@ -57,14 +57,24 @@ else
...
 endif
 
-CFLAGS += ${DEFINE_PREFIX}USE_PTHREAD ${DEFINE_PREFIX}PERFORMANCE_RUN=1
-CFLAGS += ${DEFINE_PREFIX}MULTITHREAD=$(COREMARK_NTHREADS)
+CFLAGS += ${DEFINE_PREFIX}CALLGRIND_RUN=0
+CFLAGS += ${DEFINE_PREFIX}COMPILER_REQUIRES_SORT_RETURN=0
+CFLAGS += ${DEFINE_PREFIX}CORE_DEBUG=$(CONFIG_COREMARK_DEBUG)
 CFLAGS += ${DEFINE_PREFIX}FLAGS_STR=\""$(ARCHOPTIMIZATION)\""
-CFLAGS += ${DEFINE_PREFIX}PRINT_ARGS=$(COREMARK_PRINT_ARGS)
 CFLAGS += ${DEFINE_PREFIX}MEM_LOCATION=\"Stack\"
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Makefile:64]:
```suggestion
CFLAGS += ${DEFINE_PREFIX}MEM_LOCATION=CONFIG_COREMARK_MEM_LOCATION
```

File: benchmarks/coremark/Kconfig [hunk 8]
```diff
@@ -27,6 +27,32 @@ config COREMARK_STACKSIZE
...
+		2: SEED_VOLATILE
+
+config COREMARK_MEM_METHOD
+	int "Coremark Memory Method"
+	default 1
+	---help---
+		Set Coremark Memory Method:
+		0: MEM_STATIC
+		1: MEM_MALLOC
+		2: MEM_STACK
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Kconfig:46]: add
```
config COREMARK_MEM_LOCATION
	string
	default "global" if COREMARK_MEM_METHOD == 0
	default "heap" if COREMARK_MEM_METHOD == 1
	default "stack" if COREMARK_MEM_METHOD == 2
```

File: benchmarks/coremark/Makefile [hunk 9]
```diff
@@ -57,14 +57,25 @@ else
...
-CFLAGS += ${DEFINE_PREFIX}USE_PTHREAD ${DEFINE_PREFIX}PERFORMANCE_RUN=1
-CFLAGS += ${DEFINE_PREFIX}MULTITHREAD=$(COREMARK_NTHREADS)
+CFLAGS += ${DEFINE_PREFIX}CALLGRIND_RUN=0
+CFLAGS += ${DEFINE_PREFIX}COMPILER_REQUIRES_SORT_RETURN=0
+CFLAGS += ${DEFINE_PREFIX}CORE_DEBUG=$(CONFIG_COREMARK_DEBUG)
 CFLAGS += ${DEFINE_PREFIX}FLAGS_STR=\""$(ARCHOPTIMIZATION)\""
+CFLAGS += ${DEFINE_PREFIX}MEM_LOCATION=$(CONFIG_COREMARK_MEM_LOCATION)
+CFLAGS += ${DEFINE_PREFIX}MEM_METHOD=$(CONFIG_COREMARK_MEM_METHOD)
+CFLAGS += ${DEFINE_PREFIX}MICA=0
+CFLAGS += ${DEFINE_PREFIX}PRINT_ARGS=0
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Makefile:67]: remove line 67 which is duplicated with line 70

File: benchmarks/coremark/Kconfig (hunk 8 above)
Comment (pengxiang-hao) [reply benchmarks/coremark/Kconfig:46]:
```
apps/benchmarks/coremark/Kconfig:50: error: couldn't parse 'default "GLOBAL" if (CONFIG_COREMARK_MEM_METHOD == 0)': expected symbol
```

it seems that `string` `default` with `equation` does not supported in nuttx Kconfig，only 
```
default "heap" if COREMARK_MEM_METHOD
```
is accepted, should I just keep the original change

```
config COREMARK_MEM_LOCATION
	string "Coremark memory location string"
	default "Defined by the NuttX configuration"
```

or using another type, such like divide these into three Y/N condition which  mapping value of `COREMARK_MEM_METHOD`?


---

```
config COREMARK_MEM_METHOD
	int "Coremark Memory Method"
	default 1
	---help---
		Set Coremark Memory Method:
		0: MEM_STATIC
		1: MEM_MALLOC
		2: MEM_STACK

config COREMARK_MEM_USE_GLOBAL
	bool "Use global memory for Coremark"
	default y if COREMARK_MEM_METHOD == 0
	---help---
	  Use global memory for Coremark

config COREMARK_MEM_USE_HEAP
	bool "Use heap memory for Coremark"
	default y if COREMARK_MEM_METHOD == 1
	---help---
	  Use heap memory for Coremark

config COREMARK_MEM_USE_STACK
	bool "Use stack memory for Coremark"
	default y if COREMARK_MEM_METHOD == 2
	---help---
	  Use stack memory for Coremark

config COREMARK_MEM_LOCATION
	string "Coremark Memory Location"
	default "Defined by the NuttX configuration" if COREMARK_MEM_USE_GLOBAL
	default "heap" if COREMARK_MEM_USE_HEAP
	default "stack" if COREMARK_MEM_USE_STACK
	---help---
	  Set Coremark Memory Location:
	  0: global
	  1: heap
	  2: stack
```

Comment (xiaoxiang781216) [reply benchmarks/coremark/Kconfig:46]: Ok, let's define MEM_LOCATION with "if/elif/endif" in Makefile.

File: benchmarks/coremark/Kconfig [hunk 10]
```diff
@@ -27,6 +27,41 @@ config COREMARK_STACKSIZE
...
+
+config COREMARK_MEM_LOCATION
```
Comment (xiaoxiang781216) [inline benchmarks/coremark/Kconfig:48]: let's remove COREMARK_MEM_LOCATION which isn't used in the new patch

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**